### ISSUE TITLE
port-check.ts does not log when port checking fails

### ISSUE THREAD

**Issue by karl-run** [testcontainers/testcontainers-node#396]
**Expected Behaviour**
Error logging telling me that the port checking is failing. In this particular case the container does not have `bash` available, so `port-check.ts` fails on checking if the port has been bound until it times out.

**Actual Behaviour**
No logs, even with `DEBUG=testcontainers*`, only times out after 60 seconds with a error message saying that the port was never bound.

**Testcontainer Logs**
```plain
  testcontainers DEBUG System diagnostics: {"node":{"version":"v16.17.0","architecture":"x64","platform":"linux"},"docker":{"serverVersion":"20.10.18","operatingSystem":"Pop!_OS 22.04 LTS","operatingSystemType":"linux","architecture":"x86_64","cpus":24,"memory":33578258432}} +1ms
  testcontainers DEBUG Not pulling image as it already exists: ghcr.io/<image> +88ms
  testcontainers DEBUG Creating new Reaper for session: f93bb52c39c6705655ea67640c7d2497 +0ms
  testcontainers DEBUG No registry auth locator found for registry: "https://index.docker.io/v1/" +0ms
  testcontainers DEBUG Checking if image exists: testcontainers/ryuk:0.3.2 +0ms
  testcontainers DEBUG Not pulling image as it already exists: testcontainers/ryuk:0.3.2 +86ms
  testcontainers INFO  Creating container for image: testcontainers/ryuk:0.3.2 +1ms
  testcontainers INFO  Starting container testcontainers/ryuk:0.3.2 with ID: d2553b4d2da9ce084fd66546ffc91589d643898b0ed7a11e7e768d8e079a0965 +62ms
  testcontainers DEBUG Waiting for container to be ready: d2553b4d2da9ce084fd66546ffc91589d643898b0ed7a11e7e768d8e079a0965 +228ms
  testcontainers DEBUG Waiting for log message "/.+ Started!/" +0ms
  testcontainers INFO  Container is ready +3ms
  testcontainers DEBUG Connecting to Reaper d2553b4d2da9ce084fd66546ffc91589d643898b0ed7a11e7e768d8e079a0965 on localhost:49263 +0ms
  testcontainers DEBUG Connected to Reaper d2553b4d2da9ce084fd66546ffc91589d643898b0ed7a11e7e768d8e079a0965 +1ms
  testcontainers INFO  Creating container for image: ghcr.io/<image> +0ms
  testcontainers INFO  Starting container ghcr.io/<image> with ID: 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465 +42ms
  testcontainers DEBUG Waiting for container to be ready: 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465 +223ms
  testcontainers DEBUG Waiting for host port 49264 for 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465 +0ms
  testcontainers DEBUG Waiting for internal port 8080 for 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465 +0ms
  testcontainers DEBUG Host port 49264 ready for 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465 +2ms
  testcontainers ERROR Container failed to be ready: Error: Port 8080 not bound after 60000ms for 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465 +1m

Error: Port 8080 not bound after 60000ms for 863d846339a565fd48a04b194840d65d0d3d9652082a5d9e5733322f77bbf465

    at /home/<dir>/node_modules/testcontainers/dist/wait-strategy.js:64:23
    at IntervalRetryStrategy.<anonymous> (/home/<dir>/node_modules/testcontainers/dist/retry-strategy.js:36:28)
    at Generator.next (<anonymous>)
    at fulfilled (/home/<dir>/node_modules/testcontainers/dist/retry-strategy.js:5:58)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

**Steps to Reproduce**
Running any container that has neither `/bin/sh` or `/bin/bash` available. You'll see that the commands in [port-check.ts:41:43](https://github.com/testcontainers/testcontainers-node/blob/master/src/port-check.ts#L41-L43) will all fail like this:

```json
[
  {
    "output": "OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory: unknown\r\n",
    "exitCode": 126
  },
  {
    "output": "OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory: unknown\r\n",
    "exitCode": 126
  },
  {
    "output": "OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/bash\": stat /bin/bash: no such file or directory: unknown\r\n",
    "exitCode": 126
  }
]
```

**Environment Information**
```json
{
  "node": { "version": "v16.17.0", "architecture": "x64", "platform": "linux" },
  "docker": {
    "serverVersion": "20.10.18",
    "operatingSystem": "Pop!_OS 22.04 LTS",
    "operatingSystemType": "linux",
    "architecture": "x86_64",
    "cpus": 24,
    "memory": 33578258432
  }
}

```

**Comment by cristianrgreco** [comment]
Hi @karl-run, thanks for raising. The reason for not logging the result of the port-check is that it can be very noisy (every 100ms we run several commands). In your example, one of the 3 commands is outputting a 126 exit code, but it doesn't mean that the other commands aren't trying with more success. I worry therefore that outputting this error is misleading.

**Comment by karl-run** [comment]
Hmm, yeah I agree that it can be noisy. But when the overall check times out, maybe one could log the result of the last attempt? That would have saved me quite a few hours of debugging today, for example. :smiling_face_with_tear:

**Comment by cristianrgreco** [comment]
@karl-run it's a good idea, I'll experiment with some approaches. I'd like to be able to fail-fast with an error message but not sure it's safe to do so. If not I'd like to output an error message ASAP to save you from waiting 2 minutes to see what failed. Worst case (if it's too noisy) we can output something at the end.

**Comment by karl-run** [comment]
It might be too specific, but `/bin/sh` not being available is becoming more and more common as [distroless](https://github.com/GoogleContainerTools/distroless) is becoming more and more common. So maybe failing fast on `/bin/sh` missing? :shrug: 

At least it would be very helpful for any poor soul in the future trying to run a distroless based image in testcontainers.

**Comment by karl-run** [comment]
Also, for any poor soul coming to this issue in the future because they are running some distroless based image, this is the solution to your problem:

```ts
.withWaitStrategy(Wait.forLogMessage(/.*some log output your container has saying it is ready.*/))
```

**Comment by cristianrgreco** [comment]
@karl-run have a look at the linked PR, specifically the test names, do you think this covers most things/is a worthwhile improvement?

**Comment by karl-run** [comment]
Looks really good! In my opinion this is definitely worth while.

**Comment by cristianrgreco** [comment]
Excellent, thanks @karl-run, released in 8.14.0

**Comment by treydavis** [comment]
I would like to keep specific container-related info out of my tests. For example, I'd like to avoid something like this:
``` js
const environment = await new DockerComposeEnvironment(
    composeFilePath,
    composeFile
).withWaitStrategy(
    'some-specific-container',
    Wait.forLogMessage(/.*some log output specific to the container.*/)
).up();
```
I'd rather create a `healthcheck` for that container in my Docker compose file so that all the container-specific configs are defined in one place. This is also an alternative solution to the problem of distroless images if you add a binary like wget as described here: https://github.com/GoogleContainerTools/distroless/issues/183#issuecomment-571723446

This approach has led me to wish for a configurable default wait strategy:
``` js
const environment = await new DockerComposeEnvironment(
    composeFilePath,
    composeFile
).withDefaultWaitStrategy(Wait.forHealthCheck());
```

Or, perhaps, an alternative signature to `withWaitStrategy` (or a separate method name):
``` typescript
const environment = await new DockerComposeEnvironment(
    composeFilePath,
    composeFile
).withWaitStrategy((container: Dockerode.Container) => {
    if (container.Labels... === "something") return Wait.forHttp(...);
    if (container.Labels... === "other") return Wait.forSomething();
    return Wait.forHealthCheck();
})
```
